lib/src/services/api.dart

```diff
@@ -1,45 +1,121 @@
 import 'dart:convert';
 import 'dart:developer';
+import 'dart:io';
 
-import 'package:http/http.dart';
-import 'package:http/io_client.dart';
+import 'package:cookie_jar/cookie_jar.dart';
+import 'package:dio/dio.dart';
+//import 'package:dio_cookie_manager/dio_cookie_manager.dart';
+import 'package:flutter_secure_storage/flutter_secure_storage.dart';
 import 'package:oneplace_illinois/src/misc/config.dart';
 import 'package:oneplace_illinois/src/misc/exceptions.dart';
+import 'package:oneplace_illinois/src/misc/cookieManager.dart';
 import 'package:oneplace_illinois/src/services/firebaseAuth.dart';
 
-class ApiService extends IOClient {
-  static final _loginUri = Uri.http(Config.baseEndpoint!, '/api/v1/user/login');
-  static const _refreshCookieName = 'refresh_jwt';
+class ApiService {
+  // NOTE: Signed cookie doesn't work with HTTP request. Only works with HTTPS.
+  static final _baseUrl = Uri.https(Config.baseEndpoint!, '');
+
+  static PersistCookieJar _cookieJar = new PersistCookieJar(ignoreExpires: false);
+  /* // use default settings for now
+  static Future<PersistCookieJar> get cookieJar async {
+    if (_cookieJar == null) {
+      Directory appDocDir = await getApplicationDocumentsDirectory();
+      String appDocPath  = appDocDir.path;
+      _cookieJar = new PersistCookieJar({
+        ignoreExpires: false, // don't save/load expired cookies
+        storage: appDocPath + '/.cookies/'
+      });
+    }
+    return _cookieJar;
+  }
+  */
+
+  final storage = new FlutterSecureStorage(); // for storing access token
 
   FirebaseAuthService _firebaseAuth;
-  String? _refreshToken;
-  String? _accessToken;
 
   ApiService({required firebaseAuth}) : this._firebaseAuth = firebaseAuth;
 
-  @override
-  Future<IOStreamedResponse> send(BaseRequest request) async {
-    if (_refreshToken == null || _accessToken == null) {
+
+  Future<Dio> getClient() async {
+    final Dio _dio = Dio();
+
+    _dio.options.baseUrl = '$_baseUrl/api/v1';
+    _dio.interceptors.clear();
+
+    _dio.interceptors.add(InterceptorsWrapper(
+      onResponse: (Response response, handler) async {
+        var data = response.data;
+
+        if (data['error'] != null) {
+          log(data.toString());
+          throw ApiException(data['error']);
+        }
+
+        // save access token (refresh token is auto-saved by CookieManager above)
+        if (data['payload']?['accessToken'] != null && data['payload']?['accessToken'] != '') {
+          await storage.write(key: 'jwt_access', value: data['payload']?['accessToken']);
+        }
+
+        return data;
+      },
+    ));
+
+    return _dio;
```

Comment:
Do you think we should just go with getClientWithAuth alone, since it's probably good to make all our API requests authenticated even if not yet necessary (maybe validation is added later)?

Reply:
Also, I think there's a way to extend the Dio client, which would make ApiService itself available for running gets, posts, etc. with the credentials automatically applied. See here: https://pub.dev/packages/dio#extends-dio-class

Reply:
If we want to use just `getClientWithAuth`, then we need to modify it to allow request even if there is no access token (for `/section/search` and `/course/search`). Extending the Dio class would be nice, but I don't see the need for it at the moment. If you are able to do it without much hassle, sure go for it.

Reply:
If we extend the class, we could pass an additional option (part of `options`) to distinguish between auth & no-auth requests.

Reply:
Yeah extending the Dio class may be as much of a hassle as extending IOClient is. Does running a `/section/search` with some query parameters and an access token cause an error?

If it does, then I suppose we will need to have separate code paths for authenticated/unauthenticated requests. I propose changing the method `getClient` to use a keyword argument `credentials` or `auth` or something that can be true or false.